ENH Use scipy.special.inv_boxcox in PowerTransformer

xuefeng-xu opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.

Since we already use `scipy.special.boxcox`, I think we can use `scipy.special.inv_boxcox` for the inverse too.
https://github.com/scipy/scipy/blob/fcf7b652bc27e47d215557bda61c84d19adc3aae/scipy/special/_boxcox.pxd#L30-L34

#### Any other comments?

The Box-Cox transformation.
![image](https://github.com/scikit-learn/scikit-learn/assets/41134380/8f280d65-77ef-4860-8b29-0da30cf9755a)


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

xuefeng-xu committed:
ENH Use scipy.special.inv_boxcox in PowerTransformer

xuefeng-xu commented:
```python
@pytest.mark.parametrize(
    "method, lmbda",
    [
        ("box-cox", 0.1),
        ("box-cox", 0.5),
        ("yeo-johnson", 0.1),
        ("yeo-johnson", 0.5),
        ("yeo-johnson", 1.0),
    ],
)
def test_optimization_power_transformer(method, lmbda):
    # Test the optimization procedure:
    # - set a predefined value for lambda
    # - apply inverse_transform to a normal dist (we get X_inv)
    # - apply fit_transform to X_inv (we get X_inv_trans)
    # - check that X_inv_trans is roughly equal to X

    rng = np.random.RandomState(0)
    n_samples = 20000
    X = rng.normal(loc=0, scale=1, size=(n_samples, 1))

    pt = PowerTransformer(method=method, standardize=False)
    pt.lambdas_ = [lmbda]
    X_inv = pt.inverse_transform(X)

    pt = PowerTransformer(method=method, standardize=False)
    X_inv_trans = pt.fit_transform(X_inv)

    assert_almost_equal(0, np.linalg.norm(X - X_inv_trans) / n_samples, decimal=2)
    assert_almost_equal(0, X_inv_trans.mean(), decimal=1)
    assert_almost_equal(1, X_inv_trans.std(), decimal=1)
```

I think we need to reconsider the test above since some values of `X` are not possible with the preset lambda.

Recall that Box-Cox transformation requires $x>0$, therefore $\lambda y + 1 = x^\lambda > 0$, where $y=(x^\lambda-1)/\lambda$. However, when $y<-2$, which is possible for random generated data, $\lambda y + 1<0$ for $\lambda=0.5$.

And the build error is during computing the inverse.
```python
lmbda = 0.5
x = -2.1
print((x * lmbda + 1) ** (1 / lmbda)) # 0.0025000000000000044 (scikit-learn's inverse of box-cox)
print(np.exp(np.log1p(lmbda * x) / lmbda)) # nan (scipy's inverse of box-cox)
```

jjerphan commented:
Sorry, I don't have any idea nor time to have a look at it at the moment. :confused:

adrinjalali commented:
@xuefeng-xu you're right regarding the test as well. It would be nice if you could fix the test for this change as well and improve the test. We can also add a test to make sure if input is not resonable, we output `nan`, and we add a bug fix entry in the changelog.

xuefeng-xu committed:
Merge remote-tracking branch 'upstream/main' into inv_boxcox

xuefeng-xu committed:
fix broken test

xuefeng-xu committed:
add test for inverse box-cox

xuefeng-xu commented:
> In the original PR #10210, `inv_boxcox` was used ([780baf7](https://github.com/scikit-learn/scikit-learn/commit/780baf7dc1dbe53a99202974e7c000dcfdd48505)) and then later removed ([df2b372](https://github.com/scikit-learn/scikit-learn/commit/df2b372b0b0980179dc0a8df907325220e11443c))

It seems that the minimum supported scipy doesn't have [special.inv_boxcox](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.inv_boxcox.html) at that time, that's why we implemeted ourself?
See the discussion between these two commits https://github.com/scikit-learn/scikit-learn/pull/6781#issuecomment-279411862 and https://github.com/scikit-learn/scikit-learn/pull/6781#discussion_r101670629

I have fixed the broken test and added a new one.

adrinjalali reviewed (approved):
LGTM. This might require a changelog entry since it's changing the behavior slightly though.

cc @glemaitre

xuefeng-xu committed:
add changelog entry in v1.6

thomasjpfan commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2312,7 +2312,7 @@ def test_power_transformer_lambda_one():
     [
         ("box-cox", 0.1),
-        ("box-cox", 0.5),
+        ("box-cox", 0.2),
```
I think the test is not generating valid data for a given value of `lambda`. Setting `lambda=0.2` resolves the issue because the generated data happens to be all less than -5.

Can we work around this by clipping `X` by `1/lambda` when the method is box-coox?

```python
X = rng.normal(loc=0, scale=1, size=(n_samples, 1))

if method == "box-cox":
    # For box-cox, means that lmbda * y + 1 > 0 or y > - 1 / lmbda
    # Clip the data here to make sure the inequality is valid.
    X = np.clip(X, - 1 / lmbda, None)
```

xuefeng-xu replied:
Thank you! Clipping the data is a good idea, but may need to add a safety number, e.g., 1e-5.
```python
X = np.clip(X, - 1 / lmbda + 1e-5, None)
```

```python
from scipy.special import inv_boxcox

lmbda = 0.5
inv_boxcox(-1/lmbda, lmbda) # 0.0 (boxcox requires strictly positive data)
inv_boxcox(-1/lmbda + 1e-5, lmbda) # 2.5e-11
```

thomasjpfan replied:
Yup adding `1e-5` makes sense to me.

thomasjpfan reviewed (commented):
Thank you for the PR @xuefeng-xu !

xuefeng-xu committed:
use clipping to make sure the data is valid

xuefeng-xu committed:
lint

thomasjpfan reviewed (approved):
LGTM
